add `TypingMode::Borrowck`

lcnr opened the pull request:
Shares the first commit with #138499, doesn't really matter which PR to land first :blush: :grin: 

Introduces `TypingMode::Borrowck` which unlike `TypingMode::Analysis`, uses the hidden type computed by HIR typeck as the initial value of opaques instead of an unconstrained infer var. This is a part of https://github.com/rust-lang/types-team/issues/129.

Using this new `TypingMode` is unfortunately a breaking change for now, see tests/ui/impl-trait/non-defining-uses/as-projection-term.rs. Using an inference variable as the initial value results in non-defining uses in the defining scope. We therefore only enable it if with `-Znext-solver=globally` or `-Ztyping-mode-borrowck`

To do that the PR contains the following changes:
- `TypeckResults::concrete_opaque_type` are already mapped to the definition of the opaque type
  - writeback now checks that the non-lifetime parameters of the opaque are universal
  - for this, `fn check_opaque_type_parameter_valid` is moved from `rustc_borrowck` to `rustc_trait_selection` 
- we add a new `query type_of_opaque_hir_typeck` which, using the same visitors as MIR typeck, attempts to merge the hidden types from HIR typeck from all defining scopes
  - done by adding a `DefiningScopeKind` flag to toggle between using borrowck and HIR typeck
  - the visitors stop checking that the MIR type matches the HIR type. This is trivial as the HIR type are now used as the initial hidden types of the opaque. This check is useful as a safeguard when not using `TypingMode::Borrowck`, but adding it to the new structure is annoying and it's not soundness critical, so I intend to not add it back.
- add a `TypingMode::Borrowck`  which behaves just like `TypingMode::Analysis` except when normalizing opaque types
   - it uses `type_of_opaque_hir_typeck(opaque)` as the initial value after replacing its regions with new inference vars
   - it uses structural lookup in the new solver

fixes #112201, fixes #132335, fixes #137751

r? @compiler-errors @oli-obk

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 674c2f468fffee57995ad240853ebdbaee946d46 with merge 7139b883a3a1be10c3c2a5591ba76ff7dcf84c78...
<!-- homu: {"type":"TryBuildStarted","head_sha":"674c2f468fffee57995ad240853ebdbaee946d46","merge_sha":"7139b883a3a1be10c3c2a5591ba76ff7dcf84c78"} -->

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13992658801/job/39179997491) [(plain)](https://github.com/rust-lang/rust/commit/674c2f468fffee57995ad240853ebdbaee946d46/checks/39179997491/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.145 Building wheels for collected packages: reuse
#12 3.146   Building wheel for reuse (pyproject.toml): started
#12 3.356   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.358   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.358   Stored in directory: /tmp/pip-ephem-wheel-cache-2khast6o/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.360 Successfully built reuse
#12 3.360 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.764 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.764 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.313 Collecting virtualenv
#12 4.369   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.629      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 16.7 MB/s eta 0:00:00
#12 4.676 Collecting distlib<1,>=0.3.7
#12 4.686   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.703      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 31.6 MB/s eta 0:00:00
#12 4.742 Collecting platformdirs<5,>=3.9.1
#12 4.752   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.796 Collecting filelock<4,>=3.12.2
#12 4.807   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.888 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.068 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.29.3
#12 5.069 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.1s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      133056 kB
DirectMap2M:     7206912 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.217
---
fmt check
fmt: checked 5925 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/ui/impl-trait/nested-hkl-lifetime.rs:28: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:48
  local time: Fri Mar 21 13:29:34 UTC 2025
  network time: Fri, 21 Mar 2025 13:29:34 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-timer commented:
Queued 7139b883a3a1be10c3c2a5591ba76ff7dcf84c78 with parent 4ac032f857b46037b55c1fc0fa702450aad37f43, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (7139b883a3a1be10c3c2a5591ba76ff7dcf84c78): [comparison URL](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78&stat=instructions:u).

### Overall result: ❌ regressions - BENCHMARK(S) FAILED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression



:exclamation: :exclamation: :exclamation: :exclamation: :exclamation: 
**Warning :warning:**: The following benchmark(s) failed to build:
- wg-grammar

:exclamation: :exclamation: :exclamation: :exclamation: :exclamation: 


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.2%, 1.3%] | 79    |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.1%, 0.8%] | 14    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.2%, 1.3%] | 79    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78&stat=max-rss)
<details>
<summary>Results (primary 0.7%, secondary -3.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.7%  | [1.4%, 2.1%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-2.2%, -2.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.7% | [-3.7%, -3.7%] | 1     |
| All ❌✅ (primary)                 | 0.7%  | [-2.2%, 2.1%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78&stat=cycles:u)
<details>
<summary>Results (secondary 2.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [1.5%, 2.3%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4ac032f857b46037b55c1fc0fa702450aad37f43&end=7139b883a3a1be10c3c2a5591ba76ff7dcf84c78&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 775.474s -> 777.431s (0.25%)
**Artifact size**: 365.53 MiB -> 365.48 MiB (-0.01%)

lcnr commented:
breakage in gll
```rust
trait ContFn {}
impl<F: FnOnce()> ContFn for F {}
fn wrap<F: ContFn>(_: impl FnOnce() -> F) -> impl ContFn {
    || ()
}

struct Rule<Pat>(Box<Rule<Pat>>, Pat);
impl<Pat> Rule<Pat> {
    fn generate_parse<'a>(&'a self) -> impl ContFn + 'a {
        move || drop(wrap(|| self.0.generate_parse()))
    }
}
```

lcnr commented:
```rust
fn prove_proj<F>(_: impl FnOnce() -> F) {}
fn recur<'a>() -> impl Sized + 'a {
    prove_proj(|| recur());
    // closure has sig `fn() -> opaque<'1>`
    // wf of `prove_proj` requires `<closure as FnOnce<()>>::Output = opaque<'2>`
    // project replaces `opaque<'2>` with its hidden type (previously an unconstrained infer var)
    // it them equates that with `opaque<'1>`
    // this previously set the hidden type of `opaque<'2>` to `opaque<'1>`
    // after member constraints `'1 == '2` and we don't consider this to be a defining use
    //
    // TODO: why does `'2` end up being universal rn
}
```

lcnr committed:
move `check_opaque_type_parameter_valid`

lcnr commented:
@rustbot ready

lcnr committed:
add `TypingMode::Borrowck`

lcnr commented on tests/ui/impl-trait/issue-55872.rs:
```diff
@@ -10,8 +10,8 @@ impl<S> Bar for S {
     fn foo<T>() -> Self::E {
-        || ()
         //~^ ERROR type parameter `T` is part of concrete type but not used in parameter list for the `impl Trait` type alias
+        || ()
```
unfortunate side-effect from HIR typeck eagerly replacing the return type with an inference var. This will be partially fixed when removing this eager normalization call. The new solver does point to the closure instead.

However, as we're still normalizing away opaques to infer vars during generalization, there may still be span regressions where we point at the place introducing the opaque instead of the place which actually constrains it. This feels unavoidable without major changes to the way we track spans for opaques/in the type system in general.

compiler-errors replied:
OK, yeah that seems like the major fallout of this PR (along w/ the bad spans for mismatched hidden types).

I guess most of these will not be a problem w/ RPIT.

lcnr commented on tests/ui/impl-trait/issue-99073-2.rs:
```diff
@@ -7,8 +7,7 @@ fn main() {
         let i: u32 = test::<i32>(-1, false);
-        //~^ ERROR concrete type differs from previous defining opaque type use
-        //~| ERROR expected generic type parameter, found `i32`
+        //~^ ERROR expected generic type parameter, found `i32`
```
error tainting :tada:

lcnr commented:
@rust-timer build d7cf428b371300a2e759d68db7f8569a6ad0cd55

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 6e4f81e34efdf5c9a3d0e58b6d40f9d0de5e1e92 with merge 423064ca6a6975340d4e353b5271613f585eda43...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6e4f81e34efdf5c9a3d0e58b6d40f9d0de5e1e92","merge_sha":"423064ca6a6975340d4e353b5271613f585eda43"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14216395098/job/39841371611)
Build commit: 423064ca6a6975340d4e353b5271613f585eda43 (`423064ca6a6975340d4e353b5271613f585eda43`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14216395098/job/39841371611"},"merge_sha":"423064ca6a6975340d4e353b5271613f585eda43"} -->

rust-timer commented:
Queued 423064ca6a6975340d4e353b5271613f585eda43 with parent c9cd7078450abc15f6b8c969b1021934fb36c2a6, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c9cd7078450abc15f6b8c969b1021934fb36c2a6&end=423064ca6a6975340d4e353b5271613f585eda43).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

lcnr committed:
rm RpitConstraintChecker

we already collect opaque types from nested items
during `mir_borrowck` of the root, checking that they
are consistent this way.

lcnr commented:
slightly worse perf, not sure why. Might be due to another variant in the `TypingMode` enum and the additional `type_of_opaque_hir_typeck` query? :thinking: this impact is acceptable to me, so I'd like to merge this as is

compiler-errors commented on tests/ui/generic-associated-types/issue-87258_a.stderr:
```diff
@@ -1,10 +1,15 @@
+LL |     fn foo<'a>() -> Self::FooFuture<'a> {
+   |        ^^^
+   |
+   = note: consider removing `#[define_opaque]` or adding an empty `#[define_opaque()]`
```
This note will need tweaking

lcnr replied:
Yeah, that's existing though and the suggestiono to add an empty `#[define_opaque()]` is applicable for ATPIT afaik?

compiler-errors replied:
Yes, the "consider removing" part should be dependent on the existence of the attr, tho.

lcnr commented on compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:
```diff
@@ -39,25 +39,14 @@ pub(super) fn find_opaque_ty_constraints_for_impl_trait_in_assoc_type(
-                }
-            }
-        }
-
```
no need to check against HIR ty with `TypingMode::Borrowck`, we already use these types as the initial value for MIR borrowck

lcnr commented on compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:
````diff
@@ -164,88 +168,39 @@ impl TaitConstraintLocator<'_> {
             return;
         }
 
-        // Calling `mir_borrowck` can lead to cycle errors through
-        // const-checking, avoid calling it if we don't have to.
-        // ```rust
-        // type Foo = impl Fn() -> usize; // when computing type for this
-        // const fn bar() -> Foo {
-        //     || 0usize
-        // }
-        // const BAZR: Foo = bar(); // we would mir-borrowck this, causing cycles
-        // // because we again need to reveal `Foo` so we can check whether the
-        // // constant does not contain interior mutability.
-        // ```
````
this is no longer relevant as we no longe rsupport non-defining uses in the defining scope and only look at functions with a `#[defines_opaque(..)]` attribute

oli-obk reviewed (commented):
makes all sense to me

compiler-errors commented on compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:
```diff
@@ -287,54 +242,42 @@ pub(super) fn find_opaque_ty_constraints_for_rpit<'tcx>(
+                hidden_ty.ty
             } else {
-                hir_opaque_ty = Some(concrete_type);
+                let hir_ty = tcx.type_of_opaque_hir_typeck(def_id).instantiate_identity();
```
Well, this is reachable if neither `-Ztyping-mode-borrowck` or `-Znext-solver=globally` are enabled, right?

lcnr replied:
We're not seeding the borrowck table anymore, we only seed the default value when encountering an opaque. If the only uses of the opaque are in dead code, we still don't have any opaque in the opaque type storage in the old solver. cc #112417

I think the new solver may actually never reach this as we should try to eagerly normalize the fn sig which defines the opaque? But for now having only uses in dead code does reach this

compiler-errors commented on compiler/rustc_middle/src/ty/typeck_results.rs:
```diff
@@ -158,7 +158,7 @@ pub struct TypeckResults<'tcx> {
     /// We also store the type here, so that the compiler can use it as a hint
```
Update comment slightly?

compiler-errors replied:
actually just fold this into ur next pr, there's no reason to block this pr

compiler-errors commented on compiler/rustc_pattern_analysis/src/rustc.rs:
```diff
@@ -135,7 +135,10 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
     /// know it.
     fn reveal_opaque_key(&self, key: OpaqueTypeKey<'tcx>) -> Option<Ty<'tcx>> {
-        self.typeck_results.concrete_opaque_types.get(&key).map(|x| x.ty)
+        self.typeck_results
```
yippee

compiler-errors reviewed (approved):
r=me,oli-obk

compiler-errors commented:
@bors r=compiler-errors,oli-obk rollup=never

bors commented:
:pushpin: Commit 509a144eed84e6ccd1227cd0d90f877a774aae70 has been approved by `compiler-errors,oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors,oli-obk 509a144eed84e6ccd1227cd0d90f877a774aae70 -->
<!-- homu: {"type":"Approved","sha":"509a144eed84e6ccd1227cd0d90f877a774aae70","approver":"compiler-errors,oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 509a144eed84e6ccd1227cd0d90f877a774aae70 with merge 17ffbc81a30c094193836a5d7f90dff273b5df93...
<!-- homu: {"type":"BuildStarted","head_sha":"509a144eed84e6ccd1227cd0d90f877a774aae70","merge_sha":"17ffbc81a30c094193836a5d7f90dff273b5df93"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14273117868/job/40017911174)
Approved by: compiler-errors,oli-obk
Pushing 17ffbc81a30c094193836a5d7f90dff273b5df93 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors,oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14273117868/job/40017911174"},"merge_sha":"17ffbc81a30c094193836a5d7f90dff273b5df93"} -->

rust-timer commented:
Finished benchmarking commit (17ffbc81a30c094193836a5d7f90dff273b5df93): [comparison URL](https://perf.rust-lang.org/compare.html?start=5337252b9952fdd9482ed6a4add17254e5bd2c40&end=17ffbc81a30c094193836a5d7f90dff273b5df93&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=5337252b9952fdd9482ed6a4add17254e5bd2c40&end=17ffbc81a30c094193836a5d7f90dff273b5df93&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5337252b9952fdd9482ed6a4add17254e5bd2c40&end=17ffbc81a30c094193836a5d7f90dff273b5df93&stat=max-rss)
<details>
<summary>Results (primary -2.3%, secondary 3.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [2.2%, 4.7%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.7%, -1.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.3% | [-2.7%, -1.9%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5337252b9952fdd9482ed6a4add17254e5bd2c40&end=17ffbc81a30c094193836a5d7f90dff273b5df93&stat=cycles:u)
<details>
<summary>Results (secondary 3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2% | [3.2%, 3.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5337252b9952fdd9482ed6a4add17254e5bd2c40&end=17ffbc81a30c094193836a5d7f90dff273b5df93&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 779.508s -> 780.342s (0.11%)
**Artifact size**: 365.96 MiB -> 366.00 MiB (0.01%)

compiler-errors commented:
@craterbot run mode=check-only start=master#5337252b9952fdd9482ed6a4add17254e5bd2c40 end=master#17ffbc81a30c094193836a5d7f90dff273b5df93
